docs: functional api (#3125)
langchain-ai/langgraph: docs/docs/concepts/functional_api.md

# Functional API

!!! warning "Experimental"
    The Functional API is currently in **beta** and is subject to change. Please report any issues or feedback to the LangGraph team.

> ## Overview
> 

Reviewer: Important IMO: We need a compelling and concise explanation for when to use the imperative API over state graph. We'll need to cite this in both conceptual and how-to guides. It's unclear to me if we are recommending the functional API for specific workflows (e.g., multi-agent) or if we imagine it someday being the primary way for developers to use LangGraph. Assuming it's the latter, maybe something like what's below?

```suggestion
## Overview

The Functional API is an alternative to [State Graph](low_level.md#stategraph) for development in LangGraph. It allows you to take advantage of LangGraph's key features for persistence, human-in-the-loop workflows, and streaming without explicitly specifying state, or control flow in terms of nodes and edges. If you are building something new with LangGraph, we recommend you try the Functional API.

```

> A **workflow** is a sequence of logic that defines the flow of an application. In LangGraph, workflows can be built using the Functional API or the Graph API.

Reviewer 2: ```suggestion
A **workflow** is a sequence of steps that defines the flow of an application. In LangGraph, workflows can be built using the Functional API or the Graph API.
```

            "action": "review",
            "essays": essay
        })
        return {
            "essay": essay,
            "review": value,
        }

    thread_id = str(uuid.uuid4())

    config = {
        "configurable": {
            "thread_id": thread_id
        }
    }

    for item in workflow.stream("cat", config):
        print(item)
    ```

    ```python
    {'write_essay': 'An essay about topic: cat'}
    {'__interrupt__': (Interrupt(value={'action': 'review', 'essays': 'An essay about topic: cat'}, resumable=True, ns=['workflow:875155a0-97c1-7dc8-c216-f37c8d6b83c1'], when='during'),)}
    ```

    An essay has been written and is ready for review. Once the review is provided, we can resume the workflow:

    ```python
    from langgraph.types import Command

    for item in workflow.stream(Command(resume="This essay is great."), config):
        print(item)
    ```

    ```pycon
    {'workflow': {'essay': 'An essay about topic: cat', 'review': 'This essay is great.'}}
    ```

    The workflow has been completed and the review has been added to the essay.

## Building Blocks

The Functional API provides two building blocks for building workflows:

> - **[Entrypoint](#entrypoint)**: Defines a workflow that can include calls to tasks, other entrypoints, or state graph nodes. Entrypoints configured with a **checkpointer** enable workflow interruption and *resumption*, allowing human-in-the-loop interactions.

Reviewer 2: need to link to a checkpointer page

standard `invoke`, `ainvoke`, `stream`, and `astream` methods.

=== "Invoke"

    ```python
    my_workflow.invoke(some_input)  # Wait for the result synchronously
    ```

=== "Async Invoke"

    ```python
    await my_workflow.ainvoke(some_input)  # Await result asynchronously
    ```

=== "Stream"
    
    ```python
    for chunk in my_workflow.stream(some_input):
        print(chunk)
    ```

=== "Async Stream"

    ```python
    async for chunk in my_workflow.stream(some_input):
        print(chunk)
    ```

### Resuming

When an entrypoint is interrupted, it can be resumed by providing the appropriate command.

=== "Invoke"

    ```python
    from langgraph.types import Command
    
    my_workflow.invoke(Command(resume=some_resume_value))
    ```

=== "Async Invoke"

    ```python
    from langgraph.types import Command
    
    await my_workflow.ainvoke(Command(resume=some_resume_value))
    ```

=== "Stream"

    ```python
    from langgraph.types import Command
    
    for chunk in my_workflow.stream(Command(resume=some_resume_value)):
        print(chunk)
    ```

=== "Async Stream"

    ```python
    from langgraph.types import Command

>     async for chunk in my_workflow.stream(Command(resume=some_resume_value)):

Reviewer 2: ```suggestion
    async for chunk in my_workflow.astream(Command(resume=some_resume_value)):
```

        print(chunk)
    ```

## Task

> A **task** represents a discrete unit of work, such as an API call or data processing step, that can be executed asynchronously. Invoking a task returns a future, which can be waited on to obtain the result.

Reviewer: ```suggestion
A **task** represents a discrete unit of work, such as an API call or data processing step, that can be executed asynchronously from within an entrypoint. Invoking a task returns a future, which can be waited on to obtain the result.
```

    ...
    return result
```

!!! important "Serialization"

    The inputs and outputs of tasks must be JSON-serializable to support checkpointing.

### Invoking a Task

Tasks can only be called from within an entrypoint, another task, or a state graph node. They **cannot** be called directly from the main application code. Calling a task produces a future-like object that can be awaited or resolved to obtain the result.

=== "Synchronous Invocation"

    ```python
    @entrypoint(checkpointer=checkpointer)
    def my_workflow(input: int) -> int:
        future = slow_computation(input)
        return future.result()  # Wait for the result synchronously
    ```

=== "Asynchronous Invocation"

    ```python
    @entrypoint(checkpointer=checkpointer)
    async def my_workflow(input: int) -> int:
        return await slow_computation(input)  # Await result asynchronously
    ```

### Lifecycle

1. **Invocation:** Calling a task returns a future-like object.
2. **Checkpointing:** Results are saved to the persistence layer.
3. **Interruption Handling:** Workflows can be paused and resumed without re-executing completed tasks.

### When to Use a Task

Use tasks when:

> - To encapsulate any source of non-determinism, such as API calls, database queries, or random number generation. This ensures that the workflow can be **resumed** after being interrupted, which is critical for **human-in-the-loop** interactions.

Reviewer: Is the fact that tasks get persisted useful outside of interrupt/resume?

Author: No, not yet -- maybe for time-travel eventually

- Work needs to be retried to handle failures or inconsistencies.
- Parallel execution is beneficial for I/O-bound tasks, allowing multiple operations to run concurrently without blocking (e.g., calling multiple APIs).

## Serialization

The inputs and outputs of `@task` and `@entrypoint` must be JSON-serializable to enable checkpointing and workflow resumption. Supported data types include dictionaries, lists, strings, numbers, and booleans.

Serialization ensures that workflow state, such as task results and intermediate values, can be reliably saved and restored. This is critical for enabling human-in-the-loop interactions, fault tolerance, and parallel execution.

Providing non-serializable inputs or outputs will result in a runtime error when a workflow is configured with a checkpointer.

## Determinism

To utilize features like **human-in-the-loop**, any randomness should be encapsulated inside of tasks. This guarantees that when a workflow is halted (e.g., for human in the loop) and then resumed, it will follow the same **sequence of steps**, even if task results are non-deterministic.

> LangGraph achieves this behavior by persisting task and sub-graph results as they execute. A well-designed workflow ensures that resuming execution follows the same sequence of steps, allowing previously computed results to be retrieved correctly without having to re-execute them. This is particularly useful for long-running tasks or tasks with non-deterministic results, as it avoids repeating previously done work and allows resuming from essentialy the same 

Reviewer 2: ```suggestion
LangGraph achieves this behavior by persisting task and sub-graph results as they execute. A well-designed workflow ensures that resuming execution follows the same sequence of steps, allowing previously computed results to be retrieved correctly without having to re-execute them. This is particularly useful for long-running tasks or tasks with non-deterministic results, as it avoids repeating previously done work and allows resuming from essentially the same 
```

        value = interrupt("question")
        return value
    ```

=== "Correct"

    In this example, the side effect is encapsulated in a task, ensuring consistent execution upon resumption.

    ```python
    from langgraph.func import task

    # Define a task 
    # highlight-next-line
    @task
    # highlight-next-line
    def write_to_file():
        with open("output.txt", "w") as f:
            f.write("Side effect executed")

    @entrypoint(checkpointer=checkpointer)
    def my_workflow(inputs: dict) -> int:
        # The side effect is now encapsulated in a task.
        write_to_file().result()
        value = interrupt("question")
        return value
    ```

## Patterns

### Parallel Execution

Tasks can be executed in parallel by invoking them concurrently and waiting for the results. This is useful for improving performance in IO bound tasks (e.g., calling APIs for LLMs).

```python
@entrypoint(checkpointer=checkpointer)
def graph(input: list[int]) -> list[str]:
    futures = [mapper(i) for i in input]
    mapped = [f.result() for f in futures]
    answer = interrupt("question")
    return [m + answer for m in mapped]
```

> Tasks can be executed in parallel by invoking them concurrently and waiting for the results. This can improve performance by leveraging multiple cores or distributed resources.

Reviewer: Duplicated text/ needs to be cleaned up

- **Resumable Graph Execution**: When graph execution may need to be **resumed** after being **interrupted** (e.g., for **human-in-the-loop**), **tasks** can encapsulate any source of non-determinism, such as API calls, database queries, or random number generation. See the [determinism](#determinism) for more details.
- **Retryable Work**: When work needs to be retried to handle failures or inconsistencies, **tasks** provide a way to encapsulate and manage the retry logic.
- **Parallel Execution**: For I/O-bound tasks, **tasks** enable parallel execution, allowing multiple operations to run concurrently without blocking (e.g., calling multiple APIs).
 
## Patterns

Below are a few simple patterns that show examples of how to use the **Functional API**.

When defining an `entrypoint`, input is restricted to the first argument of the function. To pass multiple inputs, you can use a dictionary.

```python
@entrypoint(checkpointer=checkpointer)
def my_workflow(inputs: dict) -> int:
    value = inputs["value"]
    another_value = inputs["another_value"]
    return value

my_workflow.invoke({"value": 1, "another_value": 2})  
```

### Parallel execution

> Tasks can be executed in parallel by invoking them concurrently and waiting for the results. This is useful for improving performance in IO bound tasks (e.g., calling APIs for LLMs).

Reviewer 2: ```suggestion
Tasks can be executed in parallel by invoking them concurrently and waiting for the results. This is useful for improving performance in I/O-bound tasks (e.g., calling APIs for LLMs).
```
